src/System.Linq/src/System/Linq/SequenceEqual.cs

```diff
@@ -28,9 +28,25 @@ public static bool SequenceEqual<TSource>(this IEnumerable<TSource> first, IEnum
                 throw Error.ArgumentNull(nameof(second));
             }
 
-            if (first is ICollection<TSource> firstCol && second is ICollection<TSource> secondCol && firstCol.Count != secondCol.Count)
+            if (first is ICollection<TSource> firstCol && second is ICollection<TSource> secondCol)
             {
-                return false;
+                if (firstCol.Count != secondCol.Count)
+                {
+                    return false;
+                }
+
+                if (firstCol is IReadOnlyList<TSource> firstList && secondCol is IReadOnlyList<TSource> secondList)
```

Comment:
sigh - yeah it does - I was looking at the `IImmutableList<T>` interface and assuming it wouldn't implement a method where it [just throws](https://github.com/dotnet/corefx/blob/master/src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableList_1.cs#L1031-L1035)